### ISSUE TITLE
Reduce image size to what is absolutely necessary

### ISSUE THREAD

**Issue by fmoc** [cirruslabs/macos-image-templates#71]
I'm using `tart` on a few mac mini M1 which have only a 256G SSD in them. Currently, I can run at most one VM that uses the `xcode` template, as I need to allocate at least 90G of disk space for each of them. This is because the [template script](https://github.com/cirruslabs/macos-image-templates/blob/baa16d28efa9cb476738e7134646be163215adb2/templates/xcode.pkr.hcl#L27) assigns said 90G and tart won't let me create smaller disk images than what the base VM uses. I have no way of adding more disk space to these mac minis. 

The base template assigns only 40G. In a local test I could build `xcode` image with a disk size of "only" 70G, and I think it could be reduced further. Given the fact that users can always allocate more disk space but never less, the `xcode` template should be adjusted.

**Comment by fkorotkov** [comment]
Unfortunately, tart itself is not capable in expanding disk of an image. Expanding requires SSHing and running disk utility.

Thankfully there is wonderful APFS! When you clone a 90GB image, you don't actually allocate 90GB. Instead, when you start running a cloned VM and change the disk, only the changes are written to the disk file. So you can safely have a single 90GB VM and run two clones on a host with only 256GB. 

In the very end, you can create a custom VM tailored for you workflows. These VMs are tend to be generic enough to run any workflows, we have customers with a lot of assets that need that much free disk space by default.

**Comment by fmoc** [comment]
> Unfortunately, tart itself is not capable in expanding disk of an image. Expanding requires SSHing and running disk utility.

That might have been packer then. My bad for not being precise. Still, making these images bigger is not really an issue.

I guess the main issue is that you're using these `xcode` VMs directly for CI purposes, so you want `tart` to allocate some extra space in which the build root can live.

> Thankfully there is wonderful APFS! When you clone a 90GB image, you don't actually allocate 90GB. Instead, when you start running a cloned VM and change the disk, only the changes are written to the disk file. So you can safely have a single 90GB VM and run two clones on a host with only 256GB.

This indeed works, but only when you have a single image only. This is not the case for my setup currently. I may have to build the base image myself again then (this is what I used to do, using a modified version of the `xcode` template, but I was hoping I could use the upstream one to save around 10 to 15 minutes of build time).

May I suggest an `xcode-minimal` image which can be used as a base for other images?

**Comment by fkorotkov** [comment]
You only need to build your own image once and push it to your own registry. You can also strip flutter with Android SDKs to even make it smaller and preinstall your packages for even faster builds.

This is unfortunate that macOS and Xcode require so much space but there is nothing we can do about it on our side. Saving 10-20% of the image size is not worth the effort IMO since there are more users with larger disks and bigger requirements for how much space needed for builds. Sorry about that 😪

On a side not if you are building custom solution and optimizing the hell out of disk usage, you can build a minimum image and than mount folder where your CI will build things. It will even give you an IO performance boost comparing to the VM's image.

**Comment by fmoc** [comment]
I will eventually take this route, I guess. So far, I've had packer rebuild the VMs on every host, but I also discovered `tart run --dir` recently, allowing me to make truly reusable images. Once I have some time for this, I'll set some CI job (I really don't want to build images manually, the only viable way is automation). We already have an OCI-compatible registry available. Thanks for explaining the background!

Also, thanks for providing tart in general! We recently made the switch and even though there are some rough edges (which I see are being worked on!), it's much better than the hypervisor we used before...
